oeps/oep-0038-Data-Modeling.rst

```diff
@@ -0,0 +1,414 @@
+
+Use a numeric primary key
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  It is recommended to use use :code:`BigAutoField`_.
+.. _BigAutoField: https://docs.djangoproject.com/en/2.2/ref/models/fields/#bigautofield
+
+-  Do not use composite based primary keys. Use a primary key column.
+
+Have updated and created timestamps
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  The preferred method for doing this in Open edX Django applications is to inherit the :code:`TimeStampedModel`_. class.
+.. _TimeStampedModel: https://django-model-utils.readthedocs.io/en/latest/models.html#timestampedmodel
+
+- Time should be stored in UTC time by setting :code:`USE_TZ=True`_. in your python config.
+.. _USE_TZ=True: https://docs.djangoproject.com/en/2.2/topics/i18n/timezones/#overview
+
+-  If for some reason you can not inherit from :code:`TimeStampedModel`_. use the following naming conventions:
+
+   -  Created date should be named: “created”
+
+   -  Updated date should be named: “modified”
+
+Foreign keys
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  Data should be joined using primary keys wherever possible
+
+-  Foreign keys should use a naming convention of :code:`<object_name>_id` where object name is the name of the table of the foreign key relationship.
+
+-  Do not join on things such as username, email address, or other dimensions of data that may change over time
+
+-  Do not join on PII
+
+- Joining between IDAs should be done by using a universally unique identifier (UUID)
+
+- In Django use `Attributes for fields with relations`_. to identify and link models with relationships.
+.. __Attributes for fields with relations: https://docs.djangoproject.com/en/2.2/ref/models/fields/#module-django.db.models.fields.related
+
+History for models involved with enrollments, courses and course metadata, or data involving financial payments and transactions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  History using django-simple-history_. .. _django-simple-history: https://django-simple-history.readthedocs.io/en/latest/
+
+-  Remember to `backfill history`_. for existing models.
+
+.. _backfill history: https://django-simple-history.readthedocs.io/en/latest/quick_start.html#existing-projects
+
+-  Where Django simple history is not an option, the following data should be captured:
+
+   -  Fields that were changed
+
+   -  Date & time of the change
+
+   -  The foreign key of of the user who initiated the change
+
+Use the correct data type for a column
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  Don’t use a :code:`IntegerField` when a :code:`BooleanField` would do.
+
+-  Use :code:`BigIntegerField`_. for your foreign keys
+
+-  Don’t store an Integer field as :code:`CharField`_..
+.. _CharField: https://docs.djangoproject.com/en/2.2/ref/models/fields/#charfield
+
+- Store UUID's as :code:`CharField`_. with a max lenght that matches the max length of the UUID.
```

Comment:
1. lenght => length
2. I see now that some code uses UUIDField: https://docs.djangoproject.com/en/2.2/ref/models/fields/#uuidfield . Should that be avoided?

Reply:
One of the issues with UUID Field is that when a monotonically incrementing integer is not included in the column loading data without duplicates becomes more difficult. It's possible that that UUID may shift of change. Including an integer based primary key de-risks a number of data duplication scenarios.

Reply:
I've also updated the OEP to point to using the UUID field from Django. Thanks for the tip!

Reply:
It makes sense why the PK should be an auto-incrementing int, but why not use `UUIDField` non-PK UUIDs? It's used all over Discovery, Registrar, and Credentials. AFAIU, on MySQL it's just var(32) under the hood, with some convenient validation and parsing on the Django level.

Reply:
Sorry, was posting on old version of this thread. Cool!

Reply:
So I agree with the decision that primary keys should be integers, but I'm curious about this statement:

> One of the issues with UUID Field is that when a monotonically incrementing integer is not included in the column loading data without duplicates becomes more difficult. It's possible that that UUID may shift of change.

My understanding is that UUIDs should never change, since one of their primary uses is to uniquely identify data across IDAs. If folks are ever mutating the UUID values on a row, I'd really like to understand why, because it sounds like something that will almost certainly bite us down the road. Or does this have more to do with the difficulty in doing range queries...?

Thank you.

Reply:
Sometimes Dave, people do goofy things. That said I'm happy to strike the sentence.